evap/results/views.py

```diff
@@ -23,6 +23,10 @@ def get_course_result_template_fragment_cache_key(course_id, language, can_user_
 
 def delete_template_cache(course):
     assert course.state != 'published'
+    delete_template_cache_impl(course)
+
+
+def delete_template_cache_impl(course):
     caches['results'].delete(get_course_result_template_fragment_cache_key(course.id, 'en', True))
```

Comment:
i get the intent that you wanted to keep the assert for the other calls to this method, but now we have two similar methods, where the only difference is an added assert. i wouldn't do that and just remove the assert. @He3lixxx if you disagree, lets talk about that.

Reply:
(All below is afaik, there might be something wrong in there, please correct me)

Keeping these as two methods way mainly my request. My idea is that this function is dangerous. I (as a non-evap guy) wouldn't expect that you can break stuff if you delete cache entries, I would rather expect the system to fix itself on the fly when it encounters missing cache entries (in the end, it's a cache that doesn't give integrity guarantees). However, we're doing something quite different. Published courses must always have valid cache entries, they will be created on publishing and removed on unpublishing. Non-published courses must not have cache entries. These are some very strict constraints, and if they are violated, we will probably end up in a state where every page rendering will be incorrect.

In favor of keeping this seperated:
- I favor embedding these constraints in the code in general rather than just having the knowledge in our heads. (for the sake of anyone new or anyone who forgets how exactly things were)
- I think that the whole code around caching the results is rather complex. I expect us (or atleast myself) to miss some situations where we will try to call these methods where they shouldn't be called in the future. The asserts (somewhat) guarantee no harm is done - the single call will fail, but we will not get a permanent invalid state.
- I like the division in a "safe public method" that other modules may call and a designated "unsafe private method" that may not be called from outside.

In favor of removing the assert:
- As long as every reviewer knows that the cache is a fragile thing here, we will probably not get code through code review that breaks stuff because we know we have to be careful about that.
- We can not assume that any call to the "safe" method is indeed safe as it has this assert and thus might raise an uncaught exception, so even with the "safe" method, we still need to be careful.

I'd say the last point makes merging these methods and removing the assert reasonable, so I don't have a strong opinion, but I'm still slightly in favor of keeping the assert. Anything I don't see so far?

Reply:
i think you understood the problem perfectly, and i agree with all you said.

what about this solution:

```python
def do_the_right_thing(courses): # TODO better name ;)
    assert (all_courses_are_published) or (all_courses_are_not_published)
    for course in courses:
         delete_template_cache(courses)
    if all_courses_are_published:
        warm_up_template_cache(courses)
```
this would be the only public interface. any code that wants to remove, add, or update a course in/from the cache would call this method.  this would always be safe, because it always does the right thing. however, it would include unnecessary work: in some cases we know a course has not been in the cache, the code would still attempt to remove it. also, on the call site, it's less clear what is actually the desired action, because it would always call `do_the_right_thing` and not `warm_up`, `delete`, or `update`.

...actually, having described the downsides, i think i like your solution better.

lets prefix `delete_template_cache_impl` with an underscore to fulfill the convention, and otherwise leave it as it is.